interp/src/stk_env.rs

```diff
@@ -0,0 +1,809 @@
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// let mut b = Smoosher::new();
+    /// b.set("hey", 13);
+    /// let c = Smoosher::merge(a, b); //there's no common fork point
+    /// ```
+    pub fn merge(self, other: Self) -> Self {
+        //find shared fork point; if doesn't exist, panic
+        let mut a = self;
+        let mut b = other;
+        if let Some((depth_a, depth_b)) = Smoosher::shared_fork_point(&a, &b) {
+            //smoosh [self] and [other] to right before that point
+            a = a.smoosh(depth_a - 1);
+            b = b.smoosh(depth_b - 1);
+
+            //now do merge
+
+            //get both heads of smooshed a and b
+            let mut a_head = a.head;
+            //merge a_head and b_head.
+            for (k, v) in b.head {
+                a_head.insert(k, v);
+            }
+            //now drop b
+            std::mem::drop(b.tail); //b.head already consumed above
+                                    //create a'
+            let (a_new_head, a_new_tail) = a.tail.split();
+            if let Some(mut a_new_head) = a_new_head {
+                a = Smoosher {
+                    head: a_new_head,
+                    tail: a_new_tail,
+                };
+            } else {
+                panic!("trying to merge, but [self] is empty")
+            }
+            //push_scope this new merged node onto A'
+            a.push_scope(a_head);
+            a.smoosh_once()
+        } else {
+            panic!("tried to merge Smooshers with no common fork point")
+        }
+    }
+
+    /// ```text
+    /// Returns a HashSet of references to keys bounded in the
+    /// top [levels] levels of the Smoosher.
+    /// [self.list_bound_vars(0)] returns all keys in the topmost scope
+    /// [self.list.bound_vars(1)] returns all keys in the top two scopes
+    /// Undefined behavior if levels >= (# of scopes), or [self] is empty
+    /// ```
+    /// # Example
+    /// ```
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// a.set("hello", 0);
+    /// let b = a.list_bound_vars(1);
+    /// assert_eq!(b.len(), 3);
+    /// assert!(b.contains(&"hello"));
+    /// assert!(b.contains(&"hi!"));
+    /// assert!(b.contains(&"bye"));
+    /// ```
+
+    pub fn list_bound_vars(&self, levels: u64) -> HashSet<&K> {
+        let mut tr_hs = HashSet::new();
+        //levels is at least 0, so add everything from head
+        for (k, _) in self.head.iter() {
+            tr_hs.insert(k);
+        }
+        //now iterate and only add if levels > 0
+        let mut levels = levels as i32;
+        for hm in self.tail.iter() {
+            if levels > 0 {
+                for (k, _) in hm.iter() {
+                    tr_hs.insert(k);
+                }
+            }
+            levels -= 1;
+        }
```

Comment:
You could use iterators for this, if you felt like it, using `take(levels)` and `fold`

Reply:
Seems like `fold` requires a functional closure, but we don't want to deal with a new `HashSet`for each iteration. So we will probably keep the `take(levels)` but not use `fold`